File: imu_calib_matlab/tests/test_apply_calibration.m
% test_apply_calibration
% Validate the unified inverse calibration entry point.

clear;
clc;

projectRoot = fileparts(fileparts(mfilename('fullpath')));
addpath(genpath(fullfile(projectRoot, 'src')));

[data, truth] = load_example_data('forceRegenerate', true, 'saveToMat', false);
[bg, biasInfo] = estimate_gyro_bias(data.static.gyro, 't', data.static.t, 'temp', data.static.temp); %#ok<ASGLU>
noiseStats = estimate_noise_stats(data.static.gyro, data.static.acc); %#ok<NASGU>
[Ca, ba, accInfo] = fit_acc_multi_pose(data.accPoses); %#ok<ASGLU>
[Cg, gyroInfo] = fit_gyro_C_from_angle_increment(data.gyroRuns, bg); %#ok<ASGLU>
[Kg, Mg] = split_KM(Cg); %#ok<ASGLU>
[Gg, gsensInfo] = fit_gyro_g_sensitivity(data.gsensRuns, bg, Cg); %#ok<ASGLU>

calib = struct();
calib.bg = bg;
calib.Ca = Ca;
calib.ba = ba;
calib.Cg = Cg;
calib.Gg = Gg;
calib.temp = struct('bgModel', struct('valid', false, 'message', 'No temp model'), ...
                    'baModel', struct('valid', false, 'message', 'No temp model'));
calib.gravity_magnitude = default_calib_options().acc_calibration.gravity_magnitude;

raw = struct();
raw.gyro = data.static.gyro;
raw.acc = data.static.acc;
raw.temp = data.static.temp;

corrected = apply_imu_calibration(raw, calib);

gyroMeanAfter = mean(corrected.bias_removed_gyro, 1).';
accMeanAfter = mean(corrected.acc, 1).';
accNormError = mean(abs(vecnorm(corrected.acc, 2, 2) - truth.g0));

assert(max(abs(gyroMeanAfter)) < 1e-3, ...
    'apply_imu_calibration failed to remove static gyro bias sufficiently.');
assert(norm(accMeanAfter - [0; 0; truth.g0]) < 0.3, ...
    'apply_imu_calibration produced unreasonable corrected static acceleration.');
assert(accNormError < 0.15, ...
    'apply_imu_calibration should keep corrected static acceleration close to |g|.');
assert(isfield(corrected, 'baT') && isequal(size(corrected.baT, 2), 3), ...
    'apply_imu_calibration should expose ba(T) evaluation results.');

fprintf('test_apply_calibration passed.\n');

File: imu_calib_python/tests/test_apply_calibration.py
import numpy as np

from imu_calib.calib.estimate_gyro_bias import estimate_gyro_bias
from imu_calib.calib.fit_acc_multi_pose import fit_acc_multi_pose
from imu_calib.calib.fit_gyro_C_from_angle_increment import fit_gyro_C_from_angle_increment
from imu_calib.calib.fit_gyro_g_sensitivity import fit_gyro_g_sensitivity
from imu_calib.io.load_example_data import load_example_data
from imu_calib.runtime.apply_imu_calibration import apply_imu_calibration


def test_apply_calibration_improves_static_metrics():
    data, truth, _ = load_example_data()
    bg, _ = estimate_gyro_bias(data.static.gyro, t=data.static.t, temp=data.static.temp)
    Ca, ba, _ = fit_acc_multi_pose(data.acc_poses)
    Cg, _ = fit_gyro_C_from_angle_increment(data.gyro_runs, bg)
    Gg, _ = fit_gyro_g_sensitivity(data.gsens_runs, bg, Cg)

    raw = {"gyro": data.static.gyro, "acc": data.static.acc}
    corrected = apply_imu_calibration(raw, {"bg": bg, "Ca": Ca, "ba": ba, "Cg": Cg, "Gg": Gg})

    raw_gyro_mean_norm = np.linalg.norm(data.static.gyro.mean(axis=0))
    corrected_gyro_mean_norm = np.linalg.norm(corrected["bias_removed_gyro"].mean(axis=0))
    assert corrected_gyro_mean_norm < raw_gyro_mean_norm

    raw_acc_norm_error = abs(np.linalg.norm(data.static.acc, axis=1).mean() - truth["g0"])
    corrected_acc_norm_error = abs(np.linalg.norm(corrected["acc"], axis=1).mean() - truth["g0"])
    assert corrected_acc_norm_error < raw_acc_norm_error
